### ISSUE TITLE
Sails transaction name issues with New Relic

### ISSUE THREAD

**Issue by jasonsims** [balderdashy/sails#2698]
I'm trying to dig into this more but I wanted to post it here in case any one has some insight. It seems that sails >= 0.10.5 is causing issues when teamed with new relic app monitoring. The problem is that transaction URLs are being malformed during reporting. 

For example, I'm seeing the following reported in the dashboard
![screen shot 2015-02-27 at 1 11 04 pm](https://cloud.githubusercontent.com/assets/1484253/6421292/3b055eb2-be82-11e4-9153-460421d8402e.png)

The reported routes are being repeated including garbage `*` chars for some reason. However, everything works as expected for a separate app running sails v0.10.0-rc9. 
![screen shot 2015-02-27 at 1 11 33 pm](https://cloud.githubusercontent.com/assets/1484253/6421293/4319131e-be82-11e4-8571-2c16d1ed9ade.png)

The [newrelic](https://github.com/newrelic/node-newrelic) library supports express and should be pulling everything it needs from the request but is failing to report property. My only thought is that sails performs some modifications to the request object that are not expected by this library. It sounds like a few other sails users have hit this based on [this post](https://discuss.newrelic.com/t/strange-transaction-names-in-new-relic-for-a-sails-js-app/15754) from the New Relic forums. The workaround provided does work but I believe doing so would cause you to lose some nice functionality like [router introspection](https://github.com/newrelic/node-newrelic#router-introspection).

Does anyone have any ideas on express related changes that may have happened between v0.10.0-rc9 and v0.10.5 which  could cause issues with an integration like this? 

**Environment**
sails 0.10.5
node 0.10.36
newrelic 1.17.0

**Comment by mdarveau** [comment]
Hi Jason,

I am seeing the same thing here. As I am currently trying out newrelic,
they reached out to see how were things going and I asked the question.
Here is the answer:
---
I looked at the transactions and segments in their RPM # which is 907247.
The data does seem kind of messed up, and I'm not sure why. One possibility
is that he might be running on a version of node.js that we don't support.
But the Environment section does not say what version he is using.

Unfortunately, I think he is going to have to open a support ticket. Note
that we did just release significant new version of our node.js agent which
includes better reporting of database metrics. So, you might suggest he
upgrade.
---

I haven't tried to upgrade the library yet.

Manuel

**Comment by randallmeeker** [comment]
Also have this on sails 0.10.5 & Node 0.10.32

**Comment by jasonsims** [comment]
Thanks @mdarveau! Unfortunately, I had opened a support ticket w/ new relic prior to this github issue and their response was "Sorry, we do not support the sails framework". I should be on their latest agent at v1.17.

**Comment by randallmeeker** [comment]
I was just told to look into this. 
https://github.com/newrelic/node-newrelic#newrelicsettransactionnamename

And I was going to use ```req.route.path``` but I'm trying to figure out where I should execute that in sails? Thoughts?

**Comment by jasonsims** [comment]
Yeah you can use that within some express middleware if you want. That seems more like a patch though. Since sails is based on express it should just work. I'm interested to see what specifically is breaking the newrelic integration.

**Comment by randallmeeker** [comment]
but sails has its own router so that may be the issue.

**Comment by jasonsims** [comment]
Sails has it's own router? I thought it just dynamically built express routes based on model definitions.

**Comment by Enkows** [comment]
@jasonsims you can try this:
```
sails.on('router:route', function(route) {
  if (route.req.options && route.req.options.controller) {
    newrelic.setControllerName(route.req.options.controller, route.req.options.action);
  }
})
```

**Comment by tjwebb** [comment]
> Sails has it's own router? I thought it just dynamically built express routes based on model definitions.

Yea, sails uses express underneath

**Comment by sgress454** [comment]
Looks like NewRelic found a bug in their support for Express 3--see here:
https://docs.newrelic.com/docs/release-notes/agent-release-notes/nodejs-release-notes/node-agent-1173

Are folks still seeing this issue after that patch?

**Comment by sailsbot** [comment]
Thanks for posting, @jasonsims.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
